pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	logtesting "github.com/jetstack/cert-manager/pkg/logs/testing"
+)
+
+func TestClockMetrics(t *testing.T) {
+	type testT struct {
+		expectedFmt string
+	}
+	tests := map[string]testT{
+		"clock time seconds as expected": {
+			expectedFmt: `
+  # HELP certmanager_clock_time_seconds The clock time given in seconds (from 1970/01/01 UTC).
+  # TYPE certmanager_clock_time_seconds counter
+	certmanager_clock_time_seconds %f
+	`,
+		},
+	}
+	for n, test := range tests {
+		t.Run(n, func(t *testing.T) {
+			m := New(logtesting.TestLogger{T: t})
+
+			if err := testutil.CollectAndCompare(m.clockTimeSeconds,
+				strings.NewReader(fmt.Sprintf(test.expectedFmt, float64(time.Now().Unix()))),
```

Comment:
I think this will become a flake in a very small number of cases (if the time between calling the metric and call time.Now() here falls over a second boundary).

Can we pass in a `clock.Clock` into metrics, and then use a static time for testing?

Here where we instantiate the metrics, we can pass in the clock defined above https://github.com/jetstack/cert-manager/blob/5e2a6883c1202739902ac94b5f4884152b810925/cmd/controller/app/controller.go#L218

Here is some prior art in using a fixed time clock https://github.com/jetstack/cert-manager/blob/5e2a6883c1202739902ac94b5f4884152b810925/pkg/controller/certificaterequests/sync_test.go#L51

Reply:
This obviously has a very very small chance of happening though :smile:

Reply:
Thanks for pointing in the right direction, made it easy to add :)